File: bankebank/ATMMachine.java
import java.util.ArrayList;
import java.util.Random;
import java.util.Scanner;

public class ATMMachine {

	private static class Account {
		private final String accountNumber;
		private final String firstName;
		private final String lastName;
		private String pin;
		private double balance;

		public Account(String accountNumber, String firstName, String lastName, String pin, double balance) {
			this.accountNumber = accountNumber;
			this.firstName = firstName;
			this.lastName = lastName;
			this.pin = pin;
			this.balance = balance;
		}
	}

	private final ArrayList<Account> accounts = new ArrayList<>();
	private Account currentUser = null;
	private final Scanner input = new Scanner(System.in);
	private final Random random = new Random();

	public static void main(String[] args) {
		ATMMachine app = new ATMMachine();
		app.run();
	}

	public void run() {
		while(true) {
			if (currentUser == null) {
				showMainMenu();
			}else {
				showTransactionMenu();
			}
		}
	}

	public void showMainMenu() {
		System.out.print("""
		\nWelcome to Banke Bank ATM
		1. Create Account
		2. Login
		3. Exit

		""");
		int choice = getIntInput("Enter option: ");
		switch(choice) {
			case 1 -> createAccount();
			case 2 -> login();
			case 3 -> {
				System.out.print("Thank you for using Banke Bank ATM");
				System.exit(0);
			}
			default -> System.out.print("Invalid option. Try again.");
		}
	}

	public void showTransactionMenu() {
		System.out.printf("\nWelcome, %s%n", currentUser.firstName);		
		System.out.print("""
		1. Check Balance
		2. Deposit
		3. Withdraw
		4. Transfer
		5. Change PIN
		6. Close Account
		7. Logout		
		""");
		int choice = getIntInput("Enter option: ");
		switch (choice) {
			case 1 -> checkBalance();
			case 2 -> deposit();
			case 3 -> withdraw();
			case 4 -> transfer();
			case 5 -> changePin();
			case 6 -> closeAccount();
			case 7 -> logout();
			default -> System.out.print("Invalid option. Try again.");
		}
	}

	public void createAccount() {
		System.out.println("\nAccount Creation");

		String firstName = getStringInput("Enter first name: ");
		while (!firstName.matches("[a-zA-Z]+")) {
			print("Invalid name. Only letters are allowed.");
			firstName = getStringInput("Enter first name: ");
		}
		String lastName = getStringInput("Enter last name: ");
		while (!lastName.matches("[a-zA-Z]+")) {
			print("Invalid name. Only letters are allowed.");
			lastName = getStringInput("Enter last name: ");
		}
		String pin;
		do {
			pin = getStringInput("Create 4-digit PIN: ");
			if (!pin.matches("\\d{4}")) {
				System.out.print("PIN must be exactly 4 digits");
			}
		} while (!pin.matches("\\d{4}"));
        

		String accountNumber = generateAccountNumber();
		 Account newAccount = new Account(accountNumber, firstName, lastName, pin, 0.0);
		accounts.add(newAccount);

		System.out.print("Account created successfully!");
		System.out.print("Your account number is: " + accountNumber);
		
	}

	public void login() {
		String accountNumber = getStringInput("Enter accountNumber: ");
		String pinAttempt = getStringInput("Enter PIN: ");

		for(Account account : accounts) {
			if(account.accountNumber.equals(accountNumber)) {
				if(account.pin.equals(pinAttempt)) {
					currentUser = account;
					print("Login successful!");
					return;
				}
				else {
					print("Incorrect PIN");
					return;
				}
			}
		}
		System.out.print("Account not found");
	}

	public void checkBalance() {
		System.out.printf("Your balance is #%.2f", currentUser.balance);
	}

	public void deposit() {
		double amount = getPositiveAmount("Enter amount to deposit: ");
		currentUser.balance += amount;
		System.out.printf("Deposit successful. New balance: #%.2f", currentUser.balance);
	}

	public void withdraw() {
		double amount = getPositiveAmount("Enter amount to withdraw: ");

		while(amount > currentUser.balance) {
			System.out.print("Insufficient funds");
			amount = getDoubleInput("Enter amount to withdraw: ");
		}
		currentUser.balance -= amount;
		System.out.printf("Withdrawal successful. New balance: #%.2f%n", currentUser.balance);
	}

	public void transfer() {
		String recipientNumber = getStringInput("Enter recipient account number: ");
		Account recipient = findAccount(recipientNumber);

		if(recipient == null) {
			System.out.print("Recipient account not found");
			return;
		}
		if(recipient == currentUser) {
			System.out.print("Cannot transfer to yourself");
			return;
		}

		double amount = getPositiveAmount("Enter amount to transfer: ");
		while(amount > currentUser.balance) {
			System.out.print("Insuffient funds");
			amount = getPositiveAmount("Enter amount to transfer: ");
		}

		currentUser.balance -= amount;
		recipient.balance += amount;
		System.out.printf("Transfer successful. New balance: #%.2f%n", currentUser.balance);
	}

	public void changePin() {
		String currentPin = getStringInput("Enter current PIN: ");
		if(!currentPin.equals(currentUser.pin)) {
			System.out.print("Incorrect PIN");
			return;
		}

		String newPin;
		do{
			newPin = getStringInput("Enter new 4-digit PIN: ");
			if(newPin != ("\\d{4}") ) {
				System.out.print("PIN must be exactly 4 digits");
			}
		}while (!newPin.matches("\\d{4}"));

		currentUser.pin = newPin;
		System.out.print("PIN changed successfully");
	}

	public void closeAccount() {
		String confirmation = getStringInput("Are you sure you want to close your account? (yes/no): ");
		if(confirmation.equalsIgnoreCase("yes")) {
			accounts.remove(currentUser);
			System.out.print("Account closed successfully");
			logout();
		} else {
			System.out.print("Account closure cancelled");
		}
	}
	
	public void logout() {
		currentUser = null;
		print("Logged out successfully");
	}

	public String generateAccountNumber() {
		return String.format("%07d", random.nextInt(9000000) + 1000000);
	}

	public Account findAccount(String accountNumber) {
		for(Account account : accounts) {
			if(account.accountNumber.equals(accountNumber)) {
				return account;
			}
		}
		return null;
	}

	public double getPositiveAmount(String prompt) {
		double amount;
		do {
			amount = getDoubleInput(prompt);
			if (amount <= 0) {
				print("Amount must be positive");
			}
		} while (amount <= 0);
		return amount;
	}


	public int getIntInput(String prompt) {
		while (true) {
			try {
				System.out.print(prompt);
				return Integer.parseInt(input.nextLine().trim());
			} catch (NumberFormatException e) {
				System.out.println("Invalid input. Please enter a valid integer.");
			}
		}
	}

	public double getDoubleInput(String prompt) {
		while (true) {
			try {
				System.out.print(prompt);
				return Double.parseDouble(input.nextLine().trim());
			} catch (NumberFormatException e) {
				System.out.print("Invalid input. Please enter a valid number.");
			}
		}
	}

	public String getStringInput(String prompt) {
		System.out.print(prompt);
		return input.nextLine().trim();
	}

	public void print(String message) {
		System.out.println(message);
	}

}

File: bankebank/atmMachine.js
const readline = require('readline');

class Account {
	constructor(accountNumber, firstName, lastName, pin, balance = 0) {
		this.accountNumber = accountNumber;
		this.firstName = firstName;
		this.lastName = lastName;
		this.pin = pin;
		this.balance = balance;
	}
}

class ATMMachine {
	constructor() {
		this.accounts = [];
		this.currentUser = null;
		this.rl = readline.createInterface({
			input: process.stdin,
			output: process.stdout
		});
	}

	generateAccountNumber() {
		return 'BNK' + Math.floor(100000 + Math.random() * 900000);
	}


	findAccount(accountNumber) {
		return this.accounts.find(account => account.accountNumber === accountNumber);
	}

	async getUserInput(prompt) {
		return new Promise(resolve => {
			this.rl.question(prompt, resolve);
		});
	}

	async showMainMenu() {
		console.log('\nWelcome to Banke Bank ATM');
		console.log('1. Create Account');
		console.log('2. Login');
		console.log('3. Exit');
	
		const choice = await this.getUserInput('Enter option: ');
		switch(choice) {
		case '1':
			await this.createAccount();
		break;
		case '2':
			await this.login();
		break;
		case '3':
			console.log('Thank you for using Banke Bank ATM');
			this.rl.close();
			process.exit();
		default:
			console.log('Invalid option');
			await this.showMainMenu();
		}
	}

	async showTransactionMenu() {
		console.log(`\nWelcome, ${this.currentUser.firstName}`);
		console.log('1. Check Balance');
		console.log('2. Deposit');
		console.log('3. Withdraw');
		console.log('4. Transfer');
		console.log('5. Change PIN');
		console.log('6. Close Account');
		console.log('7. Logout');

		const choice = await this.getUserInput('Enter option: ');

		switch(choice) {
			case '1':
				await this.checkBalance(); break;
			case '2':
				await this.deposit(); break;
			case '3':
				await this.withdraw(); break;
			case '4':
				await this.transfer(); break;
			case '5':
				await this.changePin(); break;
			case '6':
				await this.closeAccount(); break;
			case '7':
				await this.logout(); return;
			default:
				console.log('Invalid option');
			}
		await this.showTransactionMenu();
	}

	async createAccount() {
		console.log('\nAccount Creation');
		const firstName = await this.getUserInput('Enter first name: ');
		const lastName = await this.getUserInput('Enter last name: ');
    
		let pin;
		do {
			pin = await this.getUserInput('Create 4-digit PIN: ');
			if (!/^\d{4}$/.test(pin)) {
				console.log('PIN must be exactly 4 digits');
			}
		} while (!/^\d{4}$/.test(pin));
    
		const accountNumber = this.generateAccountNumber();
		const newAccount = new Account(accountNumber, firstName, lastName, pin);
		this.accounts.push(newAccount);
		
		console.log('\nAccount created successfully!');
		console.log(`Your account number is: ${accountNumber}`);
	}

	async login() {
		const accountNumber = await this.getUserInput('Enter account number: ');
		const pinAttempt = await this.getUserInput('Enter PIN: ');
	
		const account = this.findAccount(accountNumber);
		if (account) {
			if (account.pin === pinAttempt) {
	 			this.currentUser = account;
	 			console.log('\nLogin successful!');
	  			await this.showTransactionMenu();
			} else {
	 			console.log('\nIncorrect PIN');
			}
		} else {
			console.log('\nAccount not found');
		}
	}

	async checkBalance() {
		console.log(`\nYour balance is: $${this.currentUser.balance.toFixed(2)}`);
	}

	async deposit() {
		const amountInput = await this.getUserInput('Enter amount to deposit: $');
		const amount = parseFloat(amountInput);
		if (isNaN(amount) || amount <= 0) {
			console.log('Amount must be positive');
			return;
		}
    
		this.currentUser.balance += amount;
		console.log(`Deposit successful. New balance: $${this.currentUser.balance.toFixed(2)}`);
	}

	async withdraw() {
		const amountInput = await this.getUserInput('Enter amount to withdraw: $');
		const amount = parseFloat(amountInput);
		if (isNaN(amount) || amount <= 0) {
			console.log('Amount must be positive');
		} else if (amount > this.currentUser.balance) {
			console.log('Insufficient funds');
		} else {
			this.currentUser.balance -= amount;
			console.log(`Withdrawal successful. New balance: $${this.currentUser.balance.toFixed(2)}`);
		}
	}

	async transfer() {
		const recipientNumber = await this.getUserInput('Enter recipient account number: ');
		if (recipientNumber === this.currentUser.accountNumber) {
			console.log('Cannot transfer to yourself');
			return;
		}
		
		const recipient = this.findAccount(recipientNumber);
		if (!recipient) {
			console.log('Recipient account not found');
			return;
		}
    
		const amountInput = await this.getUserInput('Enter amount to transfer: $');
		const amount = parseFloat(amountInput);
		if (isNaN(amount) || amount <= 0) {
			console.log('Amount must be positive');
		} else if (amount > this.currentUser.balance) {
			console.log('Insufficient funds');
		} else {
			this.currentUser.balance -= amount;
			recipient.balance += amount;
			console.log(`Transfer successful. New balance: $${this.currentUser.balance.toFixed(2)}`);
		}
	}

	async changePin() {
		const currentPin = await this.getUserInput('Enter current PIN: ');
		if (currentPin !== this.currentUser.pin) {
			console.log('Incorrect PIN');
			return;
		}
    
		let newPin;
		do {
			newPin = await this.getUserInput('Enter new 4-digit PIN: ');
			if (!/^\d{4}$/.test(newPin)) {
				console.log('PIN must be exactly 4 digits');
			}
		} while (!/^\d{4}$/.test(newPin));
	
		this.currentUser.pin = newPin;
		console.log('PIN changed successfully');
	}

	async closeAccount() {
		const confirmation = (await this.getUserInput('Are you sure you want to close your account? (yes/no): ')).toLowerCase();
		if (confirmation === 'yes') {
			this.accounts = this.accounts.filter(acc => acc !== this.currentUser);
			console.log('Account closed successfully');
			await this.logout();
		} else {
			console.log('Account closure cancelled');
		}
	}

	async logout() {
		this.currentUser = null;
		console.log('Logged out successfully');
	}

	async run() {
		while (true) {
			if (this.currentUser === null) {
				await this.showMainMenu();
			} else {
				await this.showTransactionMenu();
			}
		}
	}
}

(async () => {
  const atm = new ATMMachine();
  await atm.run();
})();